### PR TITLE
add missing autocompletions

### LINKED ISSUE #501: add missing autocompletions
**Describe the solution you'd like**
There are some auto completions from 1.6 still missing in 2.0. They are:
- [x] images
- [x] links
- [x] container (`:::info` and such)
- [x] header
- [x] externals (pdf is the only one we still actively support)
- [x] extra tags ([extra.js code](https://github.com/codimd/server/blob/2890022d4d2a8bd04cd6b8895e2e28233dcdd767/public/js/index.js#L3242))
- [x] code block
If the user starts an code block an autocompletion should indicate which languages are supprted.
Bildschirmfoto von 2020-08-29 00-39-25

### PR DESCRIPTION
**Pull request by DerMolly**
### Component/Part
editor

### Description
This PR adds the missing autocompletions.

### Steps

- [x] added implementation
- [x] added / updated tests
- [ ] added / updated documentation
- [ ] extended changelog

### Related Issue(s)
closes #501

### CODE REVIEW

File: src/components/editor/editor-pane/autocompletion/link-and-extra-tag.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import { Editor, Hint, Hints, Pos } from 'codemirror'
+import moment from 'moment'
+import { getUser } from '../../../../redux/user/methods'
+import { findWordAtCursor, Hinter } from './index'
+
+const allowedChars = /[[]]*/
+const wordRegExp = /^(\[(.*])?)$/
+const allSupportedLinks = [
+  '[link text](https://example.com)',
+  '[reference]: https://example.com',
```
Comment (ErikMichelson) [inline src/components/editor/editor-pane/autocompletion/link-and-extra-tag.ts:10]: If we would want to be equal to version 1.6:

```suggestion
  '[link text](https:// "title")',
  '[reference]: https:// "title"',
```

File: src/components/editor/editor-pane/autocompletion/image.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import { Editor, Hint, Hints, Pos } from 'codemirror'
+import { findWordAtCursor, Hinter } from './index'
+
+const allowedChars = /(!|\[|]])*/
+const wordRegExp = /^(!(\[.*])?)$/
+const allSupportedImages = [
+  '![image alt](https://example.com)',
+  '![image alt](https://example.com =WidthxHeight)',
```
Comment (ErikMichelson) [inline src/components/editor/editor-pane/autocompletion/image.ts:8]: Even if the title-attribute is currently not yet implemented, we maybe want to include it here as in 1.6?

```suggestion
  '![image alt](https:// "title")',
  '![image alt](https:// "title" =WidthxHeight)',
```

**Review by ErikMichelson — COMMENTED** [review]
All in all working fine. Some minor remarks, that I don't insist on but just wanted to point out.

**Comment by DerMolly** [comment]
Maybe it should also be changed in the link autocompletion? I mean deleting `example.com` each time you use it get tiresome quick

File: src/components/editor/editor-pane/autocompletion/link-and-extra-tag.ts (hunk 1 above)
Comment (DerMolly) [reply src/components/editor/editor-pane/autocompletion/link-and-extra-tag.ts]: done

File: src/components/editor/editor-pane/autocompletion/image.ts (hunk 2 above)
Comment (DerMolly) [reply src/components/editor/editor-pane/autocompletion/image.ts]: done

File: src/components/editor/editor-pane/autocompletion/link-and-extra-tag.ts (hunk 1 above)
Comment (ErikMichelson) [reply src/components/editor/editor-pane/autocompletion/link-and-extra-tag.ts]: It looks like you removed the example.com but didn't add the title-attribute-placeholder. Is this intentional?

Comment (DerMolly) [reply src/components/editor/editor-pane/autocompletion/link-and-extra-tag.ts]: I guess I did not realize that the link could have this too. just a moment

Comment (DerMolly) [reply src/components/editor/editor-pane/autocompletion/link-and-extra-tag.ts]: done

**Review by ErikMichelson — APPROVED** [review]

**Pull request merged**